File: blind-75/top_k_freq_words.py
# leetcode 692

from typing import List
from collections import defaultdict
from collections import deque
from heapq import heappush, heappop

class Solution:
	def topKFrequent(self, words: List[str], k: int) -> List[str]:
		
		countDict = defaultdict(int)
		wordDict = defaultdict(deque)
		maxHeap = []

		for word in words:
			countDict[word] += 1

		for key, val in countDict.items():
			wordQ = None
			if val not in wordDict:
				heappush(maxHeap, (0 - val))
				wordQ = deque()
			else:
				wordQ = wordDict[val]

			wordQ.append(key)
			wordDict[val] = wordQ

		res = []
		while k > 0:
			count = heappop(maxHeap)
			wordQ = wordDict[(0 - count)]
			wordQ = deque(sorted(wordQ, key=lambda x: x))
			while len(wordQ) > 0 and k > 0:
				res.append(wordQ.popleft())
				k -= 1

		return res

obj = Solution()
# words = ["i","love","leetcode","i","love","coding"]
# k = 2
# print(obj.topKFrequent(words, k))

def test_001():
	words = ["i","love","leetcode","i","love","coding"]
	k = 2
	assert(obj.topKFrequent(words, k)) == ["i", "love"]

def test_002():
	words = ["the","day","is","sunny","the","the","the","sunny","is","is"]
	k = 4
	assert(obj.topKFrequent(words, k)) == ["the", "is", "sunny", "day"]

def test_003():
	words = ["i","love","leetcode","i","love","coding"]
	k = 3
	assert(obj.topKFrequent(words, k)) == ["i", "love", "coding"]

File: blind-75/top_k_freq_words.java
import java.util.*;

class Solution {
	public List<String> topKFrequent(String[] words, int k) {
		Map<String, Integer> countMapper = new HashMap<>();
		Map<Integer, PriorityQueue<String>> wordMapper = new HashMap<>();
		PriorityQueue<Integer> maxHeap = new PriorityQueue<>();
		List<String> res = new ArrayList<>();

		for (String word: words) {
			countMapper.put(word, countMapper.getOrDefault(word, 0) + 1);
		}

		for (Map.Entry<String, Integer> entry : countMapper.entrySet()) {
			if (!wordMapper.containsKey(entry.getValue())) {
				maxHeap.add((0 - entry.getValue()));
			}
			PriorityQueue<String> wordQ = wordMapper.getOrDefault(entry.getValue(), new PriorityQueue<>());
			wordQ.add(entry.getKey());
			wordMapper.put(entry.getValue(), wordQ);
		}

		System.out.println(countMapper);
		System.out.println(wordMapper);

		while (k-- > 0) {
			int currCount = (0 - maxHeap.poll());
			System.out.println("currCount: " + currCount);
			Queue<String> wordQ = wordMapper.get(currCount);
			// Collections.sort(wordQ);
			System.out.println(wordQ);
			do {
				System.out.println("k is: " + k);
				System.out.println("wordQ peek " + wordQ.peek());
				res.add(wordQ.poll());
			} while (!wordQ.isEmpty() && k-- > 0);
		}

		return res;
	}
}


class Main {
	public static void main (String[] args) {
		Solution obj = new Solution();

		String[] words = new String[] {"i", "love", "leetcode", "i", "love", "coding"};
		int k = 3;
		List<String> topFreqWords = obj.topKFrequent(words, k);
		System.out.println(Arrays.toString(topFreqWords.toArray()));


		words = new String[] {"the","day","is","sunny","the","the","the","sunny","is","is"};
		k = 4;
		topFreqWords = obj.topKFrequent(words, k);
		System.out.println(Arrays.toString(topFreqWords.toArray()));
	}
}